functions/access.js

```diff
@@ -52,65 +60,95 @@ const getAccessDoc = async (venueId, method) => {
 };
 
 const isValidPassword = async (venueId, password) => {
+  if (!venueId || !password) return false;
+
   const access = await getAccessDoc(venueId, "password");
-  if (!access.exists || !access.data().password) {
+
+  if (!access || !access.exists || !access.data().password) {
     return false;
   }
+
   return passwordsMatch(access.data().password, password);
 };
 
 const isValidEmail = async (venueId, email) => {
+  if (!venueId || !email) return false;
+
   const access = await getAccessDoc(venueId, "emails");
-  if (!access.exists || !access.data().emails) {
+
+  if (!access || !access.exists || !access.data().emails) {
     return false;
   }
+
   return access.data().emails.includes(email.trim().toLowerCase());
 };
 
 const isValidCode = async (venueId, code) => {
+  if (!venueId || !code) return false;
+
   const access = getAccessDoc(venueId, "code");
-  if (!access.exists || !access.data().codes) {
+
+  if (!access || !access.exists || !access.data().codes) {
     return false;
   }
+
   return access.data().codes.includes(code.trim());
 };
 
 const createToken = async (venueId, uid, password, email, code) => {
-  if (!venueId || !uid || !password || !email || !code) return {};
+  if (!venueId || !uid || !password || !email || !code) return undefined;
 
-  const venue = await admin.firestore().collection("venues").doc(venueId).get();
-  if (!venue.exists) {
+  const venueRef = admin.firestore().collection("venues").doc(venueId);
+  const accessRef = admin.firestore().collection("accessgranted").doc(uid);
+
+  if (!venueRef.exists) {
     throw new HttpsError("not-found", `venue ${venueId} does not exist`);
   }
 
-  const token = uuid();
-  const grantedDoc = await venue.ref.collection("accessgranted").doc(uid).get();
-
-  const tokenData = {
-    usedAt: [Date.now()],
-    password: password || null,
-    email: email || null,
-    code: code || null,
-  };
-
-  if (grantedDoc.exists) {
-    await grantedDoc.update({ [`tokens.${token}`]: tokenData });
-  } else {
-    const granted = { tokens: { [token]: tokenData } };
-    await grantedDoc.set(granted);
-  }
-  return { token };
+  return await admin
+    .firestore()
+    .runTransaction(async (transaction) => {
+      const [venue, granted] = await Promise.all([
+        transaction.get(venueRef),
+        transaction.get(accessRef),
+      ]);
+
+      if (!venue.exists) {
+        return undefined;
+      }
```

Comment:
In the `venueRef.exists` code above we `throw new HttpsError("not-found"`, yet here we seem to do effectively the same check but return `undefined`. Should we be aligning these?

Reply:
This has been fixed, however needs to be handled correctly when the `createToken` fails with this error. Any suggestions when it comes to UX? I can't redirect him to the landing page because the venue doesn't exists. Maybe I can send him to the default sparkle page but it would be slightly confusing.

Reply:
Would it be sufficient to just leave them on the `/:venueId` page, and show them some kind of "venue doesn't exist" error? (which I think we currently, or at least should have some concept of in the platform?)